field/TinyLang/Field/Typed/Core.hs

```diff
@@ -197,63 +227,100 @@ instance Eq f => Eq (Expr f a) where
+freeVar :: UniVar f a -> State (ScopedVarSigs f) ()
+freeVar (UniVar uni (Var uniq name)) = do
+    ScopedVarSigs free bound <- get
+    let sig   = VarSig name uni
+        free' = insertUnique uniq sig free
+    put $ ScopedVarSigs free' bound 
+
+-- | Check if variable is tracked in bound or free variables
+isTracked :: UniVar f a -> State (ScopedVarSigs f) Bool
+isTracked (UniVar uni (Var uniq name)) = do
+    ScopedVarSigs free bound <- get
+    pure $ isTrackedIn free || isTrackedIn bound
         where
-            ScopedVarSigns free bound = signs
-            sign = VarSign name uni
-            tracked = isTracked uniq sign bound || isTracked uniq sign free
-    goExpr signs (EAppUnOp _ x) = goExpr signs x
-    goExpr signs (EAppBinOp _ x y) = goExpr (goExpr signs x) y
-    goExpr signs (EIf b x y) = goExpr (goExpr (goExpr signs b) x) y
-    goExpr signs (EStatement st e) = goExpr (goStat signs st) e
-
-exprFreeVarSigns :: Expr f a -> Env (VarSign f)
-exprFreeVarSigns = _scopedVarSignsFree . exprVarSigns
-
-supplyFromAtLeastFree :: MonadSupply m => Expr f a -> m ()
-supplyFromAtLeastFree =
-    supplyFromAtLeast . freeUniqueIntMap . unEnv . _scopedVarSignsFree . exprVarSigns
+            sig = VarSig name uni
+            isTrackedIn env =
+                case lookupUnique uniq env of
+                    Just x'
+                        | x' == sig -> True
+                        | otherwise -> error $ concat [ "panic: mismatch: '"
+                                                      , show sig
+                                                      , "' vs '"
+                                                      , show x'
+                                                      , "'"]
+                    Nothing -> False
+
+-- | Gather VarSigs for a statement
+stmtVS :: Statement f -> State (ScopedVarSigs f) ()
+stmtVS (EAssert expr)    = exprVS expr
+stmtVS (ELet uniVar def) = do
+    exprVS def
+    bindVar uniVar
+stmtVS (EFor uniVar _ _ stmts) = do
+    bindVar uniVar
+    traverse_ stmtVS stmts
+
+-- | Gather VarSigs for an expression
+exprVS :: Expr f a -> State (ScopedVarSigs f) ()
+exprVS (EConst _) = pure ()
+exprVS (EVar uniVar) = do
+    tracked <- isTracked uniVar
+    unless tracked $ freeVar uniVar
+exprVS (EAppUnOp _ x) = exprVS x
+exprVS (EAppBinOp _ x y) = do
+    exprVS x
+    exprVS y
+exprVS (EIf b x y) = do
+    exprVS b
+    exprVS x
+    exprVS y
+
+-- | Collect all bindings
+execSVS :: State (ScopedVarSigs f) () -> ScopedVarSigs f
+execSVS s = execState s $ ScopedVarSigs mempty mempty
+
+-- | Return free variable signatures for a given program
+progFreeVarSigs :: Program f -> Env (VarSig f)
+progFreeVarSigs = _scopedVarSigsFree . execSVS . traverse_ stmtVS
+
+stmtsSupplyFromAtLeastFree :: MonadSupply m => Statements f -> m ()
+stmtsSupplyFromAtLeastFree =
+    supplyFromAtLeast 
+    . freeUniqueIntMap
+    . unEnv
+    . _scopedVarSigsFree
+    . execSVS
+    . traverse_ stmtVS
+
```

Comment:
@effectfully, I've cleaned up the implementation of collecting `VarSig`s throughout the program.  What do you think?

Reply:
I spent a bit of time thinking about this whole thing and I think there's a major problem with it (both the old version and the new one): it relies on global uniqueness of variables. Otherwise `x_0` in `let x_0 = x_0;` is both bound and free, and we do have programs like that. In the compiler I rename the term, so there it's fine. In old `Generator.hs` it should also be fine as we call `adjustUniquesForVars` there. But the requirement to have globally unique variables might be troubling in the new code.

So I think we should separate this procedure into two parts:

1. find free variables of a program (with variable declarations / binding external variables in `main` this won't require traversing the entire program) -- does not rely on global uniqueness, can be used in the new code freely
2. compute string names of all variables (regardless of whether they're free or bound) -- required by the compiler, relies on global uniqueness

Reply:
Just to show off our new generators and shrinkers:  after enabling all the tests quick check found the case we are discussing and showed how it can break:
```
        ERROR
          evaluation of the renamed program did not match the evaluation of the original program

        INITIAL STATE
          Env {unEnv = fromList [(23,Some (UniConst Bool True))]}

        ORIGINAL PROGRAM
          let ?n_23 = ?n_23;

        ORIGINAL PROGRAM FINAL STATE
          Right (Env {unEnv = fromList [(23,Some (UniConst Bool True))]})

        RENAMED PROGRAM
          let ?n_24 = ?n_23;

        RENAMED PROGRAM FINAL STATE
          Right (Env {unEnv = fromList [(23,Some (UniConst Bool True)),(24,Some (UniConst Bool True))]})

        ProgramWithEnv (Program (Statements [ELet (UniVar {_uniVarUni = Bool, _uniVarVar = ?n_23}) (EVar (UniVar {_uniVarUni = Bool, _uniVarVar = ?n_23}))])) (Env {unEnv = fromList [(23,Some (UniConst Bool True))]})
        Use --quickcheck-replay=831249 to reproduce.
```

Reply:
I wouldn't call it a breakage. Yes, final states are different, but that is fine as long as the final state of the renamed program corresponds to the final state of the original program, which is what we see here.

We could try extracting the final environment of the renamer and unrename the final state of the renamed program. Not sure, if it's going to work.